src/client/event_handler.rs

```diff
@@ -14,17 +14,17 @@ macro_rules! event_handler {
     ( $(
         $( #[doc = $doc:literal] )*
         $( #[cfg(feature = $feature:literal)] )?
-        async fn $method_name:ident(&self, $variant_name:ident { $(
+        async fn $method_name:ident(&self, $($context:ident: Context,)? | $variant_name:ident { $(
```

Comment:
Will get to full review later, but why is this `|` here? Shouldn't the matching work fine without it?

Reply:
Ok, apparently not, since both `$context` and `$variant_name` are identifiers, the parsing becomes ambiguous. This could still be done without the bar by defining a `event_handler_method!` macro that generates one method at a time and is called inside `event_handler!` for each method. It could have a second rule that takes a special `@NO-CONTEXT` token at the start which then wouldn't generate a `ctx: Context` parameter.

Reply:
I'm not the best at macro_rules syntax, I'd prefer if this somewhat janky syntax is kept for now and a followup PR could correct it if possible.